nixos/modules/services/backup/borgbackup.nix

```diff
@@ -0,0 +1,566 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  isPath = x: builtins.substring 0 1 x == "/";
+ 
+  mkExcludeFile = cfg:
+    # Write each exclude pattern to a new line
+    pkgs.writeText "excludefile" (concatStringsSep "\n" cfg.exclude);
+
+  mkKeepArgs = cfg:
+    # If cfg.prune.keep e.g. has a yearly attribute,
+    # its content is passed on as --keep-yearly
+    concatStringsSep " "
+      (mapAttrsToList (x: y: "--keep-${x}=${toString y}") cfg.prune.keep);
+
+  mkBackupScript = cfg: ''
+    on_exit()
+    {
+      exit_status=$?
+      # Reset the EXIT handler, or else we're called again on 'exit' below
+      trap - EXIT
+      ${cfg.postHook}
+      exit $exit_status
+    }
+    trap 'on_exit' INT TERM QUIT EXIT
+
+    archiveName="$archiveName-$(date ${cfg.dateFormat})"
+    ${cfg.preHook}
+  '' + optionalString cfg.doInit ''
+    # Run borg init if the repo doesn't exist yet
+    if ! borg list > /dev/null; then
+      borg init \
+        --encryption ${cfg.encryption.mode} \
+        $extraInitArgs
+      ${cfg.postInit}
+    fi
+  '' + ''
+    archiveSuffix="${optionalString cfg.appendFailedSuffix ".failed"}"
+    borg create \
+      --compression ${cfg.compression} \
+      --exclude-from ${mkExcludeFile cfg} \
+      $extraCreateArgs \
+      "::$archiveName$archiveSuffix" \
+      ${escapeShellArgs cfg.paths}
+  '' + optionalString cfg.appendFailedSuffix ''
+    borg rename "::$archiveName.failed" "$archiveName"
+  '' + ''
+    ${cfg.postCreate}
+  '' + optionalString (cfg.prune.keep != { }) ''
+    borg prune \
+      ${mkKeepArgs cfg} \
+      --prefix ${escapeShellArg cfg.prune.prefix} \
+      $extraPruneArgs
+    ${cfg.postPrune}
+  '';
+
+  mkPassEnv = cfg: with cfg.encryption;
+    if passCommand != null then
+      { BORG_PASSCOMMAND = passCommand; }
+    else if passphrase != null then
+      { BORG_PASSPHRASE = passphrase; }
+    else { };
+
+  mkBackupService = name: cfg: 
+    let
+      userHome = config.users.users.${cfg.user}.home;
+    in nameValuePair "borgbackup-job-${name}" {
+      description = "BorgBackup job ${name}";
+      path = with pkgs; [
+        borgbackup openssh
+      ];
+      script = mkBackupScript cfg;
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
+        # Only run when no other process is using CPU or disk
+        CPUSchedulingPolicy = "idle";
+        IOSchedulingClass = "idle";
+        ProtectSystem = "strict";
+        ReadWritePaths =
+          [ "${userHome}/.config/borg" "${userHome}/.cache/borg" ]
+          # Borg needs write access to repo if it is not remote
+          ++ optional (isPath cfg.repo) cfg.repo;
```

Comment:
A thing I just noticed: If you specify a relative path for `cfg.repo`, this will not trigger and borg won't have access to the repo. I think I should add a fancier test for `isPath`, e.g. check if no `:` is contained. It appears like all of
```
borg init -e none foo:bar
borg init -e none foo\:bar
borg init -e none foo\\:bar
```
try to use SSH. So I don't see a possibility for using a relative path with a colon. I however did not read completely through https://github.com/borgbackup/borg/issues/1705 and https://github.com/borgbackup/borg/pull/1710.

Reply:
I'm stupid. Of course there is a way: You can use `./foo:bar`.

Reply:
That's the best I could come up with:
```
isLocalPath = x:
    builtins.substring 0 1 x == "/"      # Absolute path
    || builtins.substring 0 1 x == "."   # Relative path
    || builtins.match "[.*:.*]" == null; # not machine:path
```